### PR TITLE
Clarify and refactor tag logic related to linking/unlinking

### PR DESCRIPTION
**Pull request by demiankatz**
This PR creates multiple interface methods to break apart the overloaded $listOrId argument to destroyResourceTagsLinksForUser(). It also renames methods to use link/unlink terminology instead of add/delete for tags, since the tags themselves are not being deleted. Where not already done, impacted methods from Row/Table classes are deprecated and replaced with better options in better locations.

### CODE REVIEW

File: module/VuFind/src/VuFind/Db/Service/ResourceTagsService.php [hunk 1]
```diff
@@ -158,27 +158,64 @@ public function deleteLinksByResourceTagsIdArray(array $ids): int
...
+     * Unlink tag rows that are not associated with a favorite list for the specified resource and user.
+     *
+     * @param int|int[]|null          $resourceId ID (or array of IDs) of resource(s) to unlink (null for ALL matching
+     * resources)
+     * @param UserEntityInterface|int $userOrId   ID or entity representing user
+     * @param int|int[]|null          $tagId      ID or array of IDs of tag(s) to unlink (null for ALL matching tags)
+     *
+     * @return void
+     */
+    public function destroyNonListResourceTagsLinksForUser(
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/ResourceTagsService.php:191]: As far as I can tell, this is never used, but since the functionality previously existed, I thought it was worth putting it in the interface.

File: module/VuFind/src/VuFind/Db/Table/UserResource.php [hunk 2]
```diff
@@ -164,8 +164,12 @@ public function destroyLinks($resource_id, $user_id, $list_id = null)
         // Remove any tags associated with the links we are removing; we don't
         // want to leave orphaned tags in the resource_tags table after we have
         // cleared out favorites in user_resource!
-        $this->getDbService(ResourceTagsServiceInterface::class)
-            ->destroyResourceTagsLinksForUser($resource_id, $user_id, $list_id);
+        $resourceTagsService = $this->getDbService(ResourceTagsServiceInterface::class);
+        if ($list_id === true) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/UserResource.php:168]: I'll replace the destroyLinks method here with better named and better clarified methods in the next stage of refactoring, but I want to get this piece settled first.

File: module/VuFind/src/VuFind/Tags/TagsService.php [hunk 3]
```diff
@@ -87,11 +95,76 @@ public function parse($tags)
...
+     * @return void
+     */
+    public function linkTagToResource(
+        string $tagText,
+        ResourceEntityInterface|int $resourceOrId,
+        UserEntityInterface|int $userOrId,
+        UserListEntityInterface|int|null $listOrId = null
+    ): void {
+        $trimmedTagText = trim($tagText);
+        if (strlen($trimmedTagText) > 0) {
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Tags/TagsService.php]: Nit: I'd compare with empty string. Even if it doesn't really make a difference here, we don't need the length.

File: module/VuFind/src/VuFind/Tags/TagsService.php [hunk 4]
```diff
@@ -87,11 +95,76 @@ public function parse($tags)
...
+     */
+    public function unlinkTagFromResource(
+        string $tagText,
+        ResourceEntityInterface|int $resourceOrId,
+        UserEntityInterface|int $userOrId,
+        UserListEntityInterface|int|null $listOrId = null
+    ) {
+        $listId = $listOrId instanceof UserListEntityInterface ? $listOrId->getId() : $listOrId;
+        $tagText = trim($tagText);
+        if (!empty($tagText)) {
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Tags/TagsService.php]: Again I'd compare with an empty string.

File: module/VuFind/src/VuFind/Tags/TagsService.php [hunk 5]
```diff
@@ -87,11 +95,76 @@ public function parse($tags)
...
+            $tags = $this->getDbTable('Tags');
+            $tagIds = [];
+            foreach ($tags->getByText($tagText, false, false) as $tag) {
+                $tagIds[] = $tag->getId();
+            }
+            if (!empty($tagIds)) {
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Tags/TagsService.php]: `!empty` isn't needed here.

File: module/VuFind/src/VuFind/Controller/AbstractRecord.php [hunk 6]
```diff
@@ -242,7 +242,7 @@ public function addtagAction()
         // Save tags, if any:
         if ($tags = $this->params()->fromPost('tag')) {
             $tagsService = $this->serviceLocator->get(TagsService::class);
-            $tagsService->addTagsToRecord(
+            $tagsService->linkTagsToRecord(
                 $driver,
                 $user,
                 $tagsService->parse($tags)
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/AbstractRecord.php]: Would it make sense to allow a string of tags as parameter to linkTagsToRecord? Then the parse method wouldn't need to be called here.

File: module/VuFind/src/VuFind/Controller/AbstractRecord.php [hunk 7]
```diff
@@ -280,7 +280,7 @@ public function deletetagAction()
 
         // Save tags, if any:
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/AbstractRecord.php]: Not part of this PR, but odd comment...

**Review by EreMaijala — APPROVED** [review]
I added comments about some minor issues, but none of them would prevent merging as is.

**Comment by demiankatz** [comment]
Thanks, @EreMaijala; I've incorporated all of your suggestions, and tests are still passing.
